src/daemon/osdkeys.vala

```diff
@@ -0,0 +1,254 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright Budgie Desktop Developers
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+ namespace Budgie {
+
+	[DBus (name = "org.budgie_desktop.BudgieOSD")]
+	interface ShowOSD : GLib.Object {
+		public abstract async void ShowOSD(HashTable<string,Variant> params) throws DBusError, IOError;
+	}
+
+	/*
+	  Controls the OSD for various controllable items
+	  volume
+	  brightness
+	  caps-lock
+	  num-lock
+	 */
+	class OSDKeys : Object {
+
+		// vars for the brightness handling
+		GLib.DBusConnection conn = null;
+		uint signal_id;
+		ShowOSD osd = null;
+		double current_brightness_level = 0.0;
+
+		// vars for the caps/numlock changes
+		unowned Gdk.Keymap? map;
+		bool capslock;
+		bool numlock;
+		bool firstrun = false;
+
+		// vars for the volume changes
+		private Gvc.MixerControl? mixer;
+		private Gvc.MixerStream? stream;
+		private ulong notify_id;
+
+		// need to handle initialisation of various methods so ensure the OSD is not accidently
+		// activated when budgie-daemon is started.
+		private bool initialising;
+
+		public OSDKeys() {
+			initialising = true;
+			try {
+				osd = Bus.get_proxy_sync (BusType.SESSION, "org.budgie_desktop.BudgieOSD",
+															"/org/budgie_desktop/BudgieOSD");
+
+			} catch (Error e) {
+				stderr.printf ("%s\n", e.message);
+			}
+
+			try {
+				conn = Bus.get_sync(GLib.BusType.SESSION, null);
+			}
+			catch(IOError e) {
+				print("%s", e.message);
+			}
+
+			signal_id = conn.signal_subscribe("org.gnome.SettingsDaemon.Power",
+											  "org.freedesktop.DBus.Properties",
+											  "PropertiesChanged", null, null,
+											  DBusSignalFlags.NONE,
+											  signal_powerchanges);
+
+			mixer = new Gvc.MixerControl("BD Volume Mixer");
+			mixer.state_changed.connect(on_mixer_state_change);
+			mixer.default_sink_changed.connect(on_mixer_sink_changed);
+			mixer.open();
+
+			// wait a short while to allow the async mixer methods to complete otherwise
+			// we see the volume OSD on startup accidently when keymap changes are invoked
+			Timeout.add(200, () => {
+				initialising = false;
+				map = Gdk.Keymap.get_for_display(Gdk.Display.get_default());
+				map.state_changed.connect(on_keymap_state_changed);
+				return false;
+			});
```

Comment:
Could we do this in the `on_mixer_state_change` handler once the mixer says it's ready?

Reply:
'fraid not.  Anything, whether the panel volume controller or any autostart file that touches the volume stuff can poke at the mixer to show the volume OSD.  

The slight delay here mitigates this effect.